coderd/updatecheck/updatecheck.go

```diff
@@ -0,0 +1,229 @@
+// Package updatecheck provides a mechanism for periodically checking
+// for updates to Coder.
+//
+// The update check is performed by querying the GitHub API for the
+// latest release of Coder.
+package updatecheck
+
+import (
+	"context"
+	"database/sql"
+	"encoding/json"
+	"io"
+	"net/http"
+	"time"
+
+	"github.com/google/go-github/v43/github"
+	"golang.org/x/mod/semver"
+	"golang.org/x/xerrors"
+
+	"cdr.dev/slog"
+
+	"github.com/coder/coder/coderd/database"
+)
+
+const (
+	// defaultURL defines the URL to check for the latest version of Coder.
+	defaultURL = "https://api.github.com/repos/coder/coder/releases/latest"
+)
+
+// Checker is responsible for periodically checking for updates.
+type Checker struct {
```

Comment:
Instead of all replicas doing the update check, could you lock the config table row so only one replica will attempt to do an update check at a period of ~24 hours?

Reply:
Do you have a suggestion for the how to do the lock or do we have a previous use-case for it? I'm not sure how to implement it with `sqlc`.

Essentially I'm imagining, start transaction => lock table in row exclusive mode => select row => (if it's time to update, fetch => update) => commit. But alas, sqlc.. 😄

Reply:
Postgres advisory locks are very easy to use, you just select the lock and it'll automatically get released when either the transaction ends or the connection ends.

```sql
SELECT pg_advisory_lock("some_name"); -- returns an error if already locked (I think)
-- or
SELECT pg_try_advisory_lock("some_name"); -- returns true if successfully locked

SELECT pg_advisory_unlock("some_name");
```

Reply:
Here's an example from v1: https://github.com/coder/v1/blob/7259332777bae82e5bef3fa2afacb63f5606306d/product/coder/pkg/dlock/pglock/pglock.go

Reply:
use `pg_advisory_xact_lock` for transactions for it to be auto-released if the transaction ends***

Reply:
I haven't used those before. Good suggestion, it looks promising but I don't see us having any precedent for using these. I guess there's always a first time 😄.

Reply:
They work well for our use in v1.

Reply:
> Instead of all replicas doing the update check, could you lock the config table row so only one replica will attempt to do an update check at a period of ~24 hours?

I might be blind here, but what exactly are the penalties? It doesn't seem to be frequent db access.

Reply:
Seems like a waste to do one update check for each replica if only one result will ever be stored and it's very easy to wrap in a transaction with a lock

Reply:
I'll defer this to a follow-up PR, this is a convenience fix more than anything. Ultimately I'd like for us to have a plan for how to deal with replicas, we can obviously have it be the wild west where the quickest draw wins, but assigning/limiting responsibilities via primary/secondary seems logical IMO.